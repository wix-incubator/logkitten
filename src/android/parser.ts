import { Parser } from '../types';
import { LineBuffer } from '../utils';
import { AndroidEntry } from './types';
import { areEntriesComingFromTheSameSource, PriorityUtils } from './utils';

export class AndroidParser {
  private _buffer = new LineBuffer();
  // In fact, this is: [] | [AndroidEntry]
  private _pendingEntry: AndroidEntry[] = [];

  private static readonly logcatRegex: RegExp =
    //   epoch        uid     pid     tid     level   tag     message
    /^\s*(\d+\.\d+)\s+(\S+)\s+(\d+)\s+(\d+)\s+(\w)\s+(.+?)\s*:\s*(.*)$/m;

  static create(): Parser<AndroidEntry> {
    const parser = new AndroidParser();
    return parser.parse.bind(parser);
  }

  parse(raw: string | Buffer, stderr: boolean): AndroidEntry[] {
    if (stderr) {
      return [];
    }

    const str = typeof raw === 'string' ? raw : raw.toString('utf-8');
    const lines = this._buffer.processChunk(str);
    return this._parseMessages(lines);
  }

  private _parseMessages(messages: string[]): AndroidEntry[] {
    const newEntries = messages.map(
      (rawMessage: string): AndroidEntry | null => {
        const match = rawMessage.match(AndroidParser.logcatRegex);
        if (match) {
          const [, epoch, uid, pid, tid, level, tag, message] = match;

          return {
            ts: Math.floor(parseFloat(epoch) * 1000),
            pid: Number.parseInt(pid, 10),
            tid: Number.parseInt(tid, 10),
            level: PriorityUtils.fromLetter(level),
            tag: tag.trim() || 'unknown',
            msg: message.trim(),
            uid,
          };
        }

        return null;
      }
    );

    const result = newEntries.reduce((acc, entry) => {
      if (!entry) {
        return acc;
      }

      const last = acc.at(-1);
      if (last && areEntriesComingFromTheSameSource(last, entry)) {
        last.msg += '\n' + entry.msg;
        return acc;
      }

      acc.push(entry);

      return acc;
    }, this._pendingEntry);

    this._pendingEntry = result.splice(-1);
    return result;
  }
}
