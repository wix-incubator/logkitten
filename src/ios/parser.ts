import { LineBuffer } from '../utils';
import { Parser } from '../types';
import type { IosEntry, RawEntryIOS } from './types';
import { PriorityUtils } from './utils';

const IGNORED_STDERR_PATTERN = /\bgetpwuid_r\b/;

export class IosParser {
  private _buffer = new LineBuffer();

  constructor() {
    this.parse = this.parse.bind(this);
  }

  static create(): Parser<IosEntry> {
    const parser = new IosParser();
    return parser.parse.bind(parser);
  }

  parse(raw: string | Buffer, stderr: boolean): IosEntry[] {
    const str = typeof raw === 'string' ? raw : raw.toString('utf-8');
    const lines = this._buffer.processChunk(str);
    return lines.flatMap((line) => this._parseLine(line, stderr));
  }

  private _parseLine(line: string, stderr: boolean): IosEntry[] {
    if (stderr) {
      return this._parseStderr(line);
    }

    let rawEntry: RawEntryIOS;
    try {
      rawEntry = JSON.parse(line);
    } catch {
      return [];
    }

    const entry: IosEntry = {
      msg: rawEntry.eventMessage || '',
      ts: Date.parse(rawEntry.timestamp),
      pid: rawEntry.processID,
      tid: rawEntry.threadID,
      level: PriorityUtils.fromName(rawEntry.messageType),
      // Platform specific fields
      category: rawEntry.category || '',
      formatString: rawEntry.formatString || '',
      processImagePath: rawEntry.processImagePath || '',
      subsystem: rawEntry.subsystem || '',
      userID: rawEntry.userID,
    };

    return [entry];
  }

  private _parseStderr(line: string): IosEntry[] {
    if (line.includes('No devices are booted.')) {
      throw new Error('No simulators are booted.');
    }

    if (IGNORED_STDERR_PATTERN.test(line)) {
      return [];
    }

    throw new Error(line);
  }
}
