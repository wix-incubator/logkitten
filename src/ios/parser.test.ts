import { Chunker } from '../__tests__/Chunker';
import { IosParser } from './parser';
import type { IosEntry } from './types';

const DUMP =
  [
    '{"timezoneName":"","messageType":"Default","eventType":"logEvent","source":null,"formatString":"Task <%{public,uuid_t}.16P>.<%lu> summary for %{public}s {transaction_duration_ms=%u, response_status=%ld, connection=%llu, reused=1, reused_after_ms=%u, request_start_ms=%u, request_duration_ms=%u, response_start_ms=%u, response_duration_ms=%u, request_bytes=%lld, request_throughput_kbps=%lld, response_bytes=%lld, response_throughput_kbps=%lld, cache_hit=%{bool}d}","userID":501,"activityIdentifier":0,"subsystem":"com.apple.CFNetwork","category":"Summary","threadID":154653588,"senderImageUUID":"28F30CFF-18C1-3FB9-B1E9-E185E004AC74","backtrace":{"frames":[{"imageOffset":265776,"imageUUID":"28F30CFF-18C1-3FB9-B1E9-E185E004AC74"}]},"bootUUID":"","processImagePath":"/Users/yaroslavs/Library/Developer/CoreSimulator/Devices/1C62AC87-73A4-4576-B440-DCE6F2A17B80/data/Containers/Bundle/Application/CEA9E514-252D-4EE2-9BA8-2E49AE389879/example.app/example","senderImagePath":"/Library/Developer/CoreSimulator/Volumes/iOS_22C150/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS 18.2.simruntime/Contents/Resources/RuntimeRoot/System/Library/Frameworks/CFNetwork.framework/CFNetwork","timestamp":"2025-06-24 14:32:34.176682+0300","machTimestamp":79751022531929,"eventMessage":"Task <96F5582A-8C99-4035-AEC1-29C47714BAD5>.<2408> summary for task failure {transaction_duration_ms=12, response_status=-1, connection=2408, reused=1, reused_after_ms=0, request_start_ms=0, request_duration_ms=0, response_start_ms=0, response_duration_ms=0, request_bytes=0, request_throughput_kbps=0, response_bytes=0, response_throughput_kbps=0, cache_hit=false}","processImageUUID":"CB3E2FA6-7D71-315B-AF0B-6AC3D30F07FE","traceID":338696976352935940,"processID":87578,"senderProgramCounter":265776,"parentActivityIdentifier":0}',
    '{"timezoneName":"","messageType":"Info","eventType":"logEvent","source":null,"formatString":"[C%u %{public}s %{public}@] dealloc","userID":501,"activityIdentifier":0,"subsystem":"com.apple.network","category":"connection","threadID":154652000,"senderImageUUID":"2CB8C5CD-213F-3259-9533-78B2684F355E","backtrace":{"frames":[{"imageOffset":4176148,"imageUUID":"2CB8C5CD-213F-3259-9533-78B2684F355E"}]},"bootUUID":"","processImagePath":"/Users/yaroslavs/Library/Developer/CoreSimulator/Devices/1C62AC87-73A4-4576-B440-DCE6F2A17B80/data/Containers/Bundle/Application/CEA9E514-252D-4EE2-9BA8-2E49AE389879/example.app/example","senderImagePath":"/Library/Developer/CoreSimulator/Volumes/iOS_22C150/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS 18.2.simruntime/Contents/Resources/RuntimeRoot/System/Library/Frameworks/Network.framework/Network","timestamp":"2025-06-24 14:32:57.470977+0300","machTimestamp":79751581592830,"eventMessage":"[C2430 Hostname#65c1aeb9:63995 tcp, url: http://localhost:63995/, definite, attribution: developer] dealloc","processImageUUID":"CB3E2FA6-7D71-315B-AF0B-6AC3D30F07FE","traceID":613355255849287940,"processID":87578,"senderProgramCounter":4176148,"parentActivityIdentifier":0}',
    '{"timezoneName":"","messageType":"Error","eventType":"logEvent","source":null,"formatString":"%{public}s %s Socket SO_ERROR %{darwin.errno}d","userID":501,"activityIdentifier":0,"subsystem":"com.apple.network","category":"connection","threadID":154652000,"senderImageUUID":"2CB8C5CD-213F-3259-9533-78B2684F355E","backtrace":{"frames":[{"imageOffset":9088656,"imageUUID":"2CB8C5CD-213F-3259-9533-78B2684F355E"}]},"bootUUID":"","processImagePath":"/Users/yaroslavs/Library/Developer/CoreSimulator/Devices/1C62AC87-73A4-4576-B440-DCE6F2A17B80/data/Containers/Bundle/Application/CEA9E514-252D-4EE2-9BA8-2E49AE389879/example.app/example","senderImagePath":"/Library/Developer/CoreSimulator/Volumes/iOS_22C150/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS 18.2.simruntime/Contents/Resources/RuntimeRoot/System/Library/Frameworks/Network.framework/Network","timestamp":"2025-06-24 14:44:49.382837+0300","machTimestamp":79768667411910,"eventMessage":"nw_socket_handle_socket_event [C3103.1.1:2] Socket SO_ERROR [61: Connection refused]","processImageUUID":"CB3E2FA6-7D71-315B-AF0B-6AC3D30F07FE","traceID":615501266340286468,"processID":87578,"senderProgramCounter":9088656,"parentActivityIdentifier":0}',
    '{"timezoneName":"","messageType":"Debug","eventType":"logEvent","source":null,"formatString":"No threshold for %{public}s, returning NW_ACTIVITY_DURATION_INVALID","userID":501,"activityIdentifier":0,"subsystem":"com.apple.network","category":"activity","threadID":154652005,"senderImageUUID":"2CB8C5CD-213F-3259-9533-78B2684F355E","backtrace":{"frames":[{"imageOffset":8408124,"imageUUID":"2CB8C5CD-213F-3259-9533-78B2684F355E"}]},"bootUUID":"","processImagePath":"/Users/yaroslavs/Library/Developer/CoreSimulator/Devices/1C62AC87-73A4-4576-B440-DCE6F2A17B80/data/Containers/Bundle/Application/CEA9E514-252D-4EE2-9BA8-2E49AE389879/example.app/example","senderImagePath":"/Library/Developer/CoreSimulator/Volumes/iOS_22C150/Library/Developer/CoreSimulator/Profiles/Runtimes/iOS 18.2.simruntime/Contents/Resources/RuntimeRoot/System/Library/Frameworks/Network.framework/Network","timestamp":"2025-06-24 14:33:21.817334+0300","machTimestamp":79752165903163,"eventMessage":"No threshold for cfnetwork:foreground_task, returning NW_ACTIVITY_DURATION_INVALID","processImageUUID":"CB3E2FA6-7D71-315B-AF0B-6AC3D30F07FE","traceID":615159717639225860,"processID":87578,"senderProgramCounter":8408124,"parentActivityIdentifier":0}',
  ].join('\n') + '\n';

describe('IosParser chunked input', () => {
  it('parses entries correctly with chunked input', () => {
    const parser = new IosParser();
    // Use the first fixture as a sample
    const chunker = new Chunker(DUMP, 4);
    let entries: IosEntry[] = [];
    for (const chunk of chunker.getChunks()) {
      entries = entries.concat(parser.parse(chunk, false));
    }
    expect(entries).toMatchSnapshot();
  });

  it('handles stderr appropriately', () => {
    const parser = new IosParser();

    // Should throw specific error for no devices
    expect(() => parser.parse('No devices are booted.\n', true)).toThrow(
      'No simulators are booted.'
    );

    // Should ignore uid errors silently
    const entries = parser.parse(
      'getpwuid_r did not find a match for uid\n',
      true
    );
    expect(entries).toEqual([]);

    // Should throw for unknown errors
    expect(() => parser.parse('Some random error\n', true)).toThrow(
      'Some random error'
    );
  });
});
