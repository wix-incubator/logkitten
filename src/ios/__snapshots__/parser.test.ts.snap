// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`IosParser chunked input parses entries correctly with chunked input 1`] = `
[
  {
    "category": "Summary",
    "formatString": "Task <%{public,uuid_t}.16P>.<%lu> summary for %{public}s {transaction_duration_ms=%u, response_status=%ld, connection=%llu, reused=1, reused_after_ms=%u, request_start_ms=%u, request_duration_ms=%u, response_start_ms=%u, response_duration_ms=%u, request_bytes=%lld, request_throughput_kbps=%lld, response_bytes=%lld, response_throughput_kbps=%lld, cache_hit=%{bool}d}",
    "level": 30,
    "msg": "Task <96F5582A-8C99-4035-AEC1-29C47714BAD5>.<2408> summary for task failure {transaction_duration_ms=12, response_status=-1, connection=2408, reused=1, reused_after_ms=0, request_start_ms=0, request_duration_ms=0, response_start_ms=0, response_duration_ms=0, request_bytes=0, request_throughput_kbps=0, response_bytes=0, response_throughput_kbps=0, cache_hit=false}",
    "pid": 87578,
    "processImagePath": "/Users/yaroslavs/Library/Developer/CoreSimulator/Devices/1C62AC87-73A4-4576-B440-DCE6F2A17B80/data/Containers/Bundle/Application/CEA9E514-252D-4EE2-9BA8-2E49AE389879/example.app/example",
    "subsystem": "com.apple.CFNetwork",
    "tid": 154653588,
    "ts": 1750764754176,
    "userID": 501,
  },
  {
    "category": "connection",
    "formatString": "[C%u %{public}s %{public}@] dealloc",
    "level": 20,
    "msg": "[C2430 Hostname#65c1aeb9:63995 tcp, url: http://localhost:63995/, definite, attribution: developer] dealloc",
    "pid": 87578,
    "processImagePath": "/Users/yaroslavs/Library/Developer/CoreSimulator/Devices/1C62AC87-73A4-4576-B440-DCE6F2A17B80/data/Containers/Bundle/Application/CEA9E514-252D-4EE2-9BA8-2E49AE389879/example.app/example",
    "subsystem": "com.apple.network",
    "tid": 154652000,
    "ts": 1750764777470,
    "userID": 501,
  },
  {
    "category": "connection",
    "formatString": "%{public}s %s Socket SO_ERROR %{darwin.errno}d",
    "level": 50,
    "msg": "nw_socket_handle_socket_event [C3103.1.1:2] Socket SO_ERROR [61: Connection refused]",
    "pid": 87578,
    "processImagePath": "/Users/yaroslavs/Library/Developer/CoreSimulator/Devices/1C62AC87-73A4-4576-B440-DCE6F2A17B80/data/Containers/Bundle/Application/CEA9E514-252D-4EE2-9BA8-2E49AE389879/example.app/example",
    "subsystem": "com.apple.network",
    "tid": 154652000,
    "ts": 1750765489382,
    "userID": 501,
  },
  {
    "category": "activity",
    "formatString": "No threshold for %{public}s, returning NW_ACTIVITY_DURATION_INVALID",
    "level": 10,
    "msg": "No threshold for cfnetwork:foreground_task, returning NW_ACTIVITY_DURATION_INVALID",
    "pid": 87578,
    "processImagePath": "/Users/yaroslavs/Library/Developer/CoreSimulator/Devices/1C62AC87-73A4-4576-B440-DCE6F2A17B80/data/Containers/Bundle/Application/CEA9E514-252D-4EE2-9BA8-2E49AE389879/example.app/example",
    "subsystem": "com.apple.network",
    "tid": 154652005,
    "ts": 1750764801817,
    "userID": 501,
  },
]
`;
